minimal_memory_lock/minimal_memory_lock.cpp

```diff
@@ -0,0 +1,273 @@
+  size_t count_;
+};
+
+int main(int argc, char * argv[])
+{
+  // Force flush of the stdout buffer.
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+
+  // Argument count and usage
+  if (rcutils_cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  // Configuration variables
+  size_t max_process_memory = 0;
+  size_t stack_size = 0;
+  bool lock_memory = false;
+  bool use_sleep = false;
+
+  // Optional argument parsing
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_LOCK_MEMORY)) {
+    lock_memory = true;
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_MEMORY_PREALLOCATION)) {
+    max_process_memory = 1024 * 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_MEMORY_PREALLOCATION));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_THREAD_STACK_SIZE)) {
+    stack_size = 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_THREAD_STACK_SIZE));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_USE_SLEEP)) {
+    use_sleep = true;
+  }
+
+  // Change a new default thread stack size to reduce memory foot-print
+  if (stack_size > 0) {
+    set_default_thread_stacksize(stack_size);
+  }
+
+  print_process_memory("Process memory before node creation: ");
+
+  // Create nodes
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<MinimalPublisher>();
+
+  // Sleep required to avoid some initial minor page faults
+  // TODO(carlos): find out why we see minor page faults initially
+  // This could be caused by middleware threads being created concurrently
+  if (use_sleep) {
+    std::cout << "Sleeping here" << std::endl;
+    std::this_thread::sleep_for(500ms);
+  }
+
+  if (lock_memory) {
+    print_process_memory("Process memory before locking: ");
+    bool memory_is_locked = configure_malloc_behavior();
+    if (memory_is_locked) {
+      std::cout << "Memory locked" << std::endl;
+    } else {
+      return EXIT_FAILURE;
+    }
+    print_process_memory("Process memory after locking: ");
+  }
+  if (max_process_memory > 0) {
+    std::cout << "Memory to reserve: " <<
+      std::to_string(max_process_memory / (1024 * 1024)) << " MB" << std::endl;
+    auto current_process_memory = get_process_memory() * 1024;
+    if (current_process_memory > max_process_memory) {
+      std::cout << "Process memory higher than maximum memory to reserve: " << std::endl;
+      return EXIT_FAILURE;
+    }
+    auto additional_memory_to_reserve = max_process_memory - current_process_memory;
+    bool memory_is_reserved = reserve_process_memory(additional_memory_to_reserve);
+    if (!memory_is_reserved) {
+      return EXIT_FAILURE;
+    }
+    print_process_memory("Process memory after memory pool reservation: ");
+  }
+
+  print_process_memory("Process memory before spin: ");
+  print_page_faults("Total page faults before spin");
+  get_new_page_faults();  // init page fault count
```

Comment:
Why is this function called and the return value not consumed?

Reply:
It's just to initialize the static members from the functions. Since this is not obvious I will change this. I can create a `MemoryPageFaultCount` class with reset and get functions to make this cleaner.

Reply:
Minor nitpick: Is it needed to store the pagefault as a state? I just store the page fault count in local variable (A) at one point in the code, do some stuff, and then get the counts again in another local variable (B). I then subtract B - A. Like https://github.com/shuhaowu/rt-demo/blob/7116d52/docs/prefault-experiments/prefault-test.cc#L80

Reply:
I'm checking the page faults inside the node callback so I need to save them either as part of the Node class or somewhere else. I started using the same approach as the rt wiki, so this is why I was using storing them inside the function itself.

Reply:
Ah maybe what you have is fine. Maybe add a comment to call out the fact that you're initialization?

Reply:
I will create a class so it's easier to understand.